### PR TITLE
Add multiplex sandboxing support to JavaBuilder

### PR DESCRIPTION
**Pull request by fmeum**
This does not include adding support to Java actions yet.

Work towards #21091

### CODE REVIEW

**Comment by fmeum** [comment]
I tried adding a test based on `--modify_execution_info`, but this doesn't quite work: The flag sets an empty string (https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/analysis/config/ExecutionInfoModifier.java;l=113), but a `"1"` is expected (https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/actions/Spawns.java;l=109).

**Review by cushon — APPROVED** [review]
Thanks! The JavaBuilder changes look great to me.

@wilwell do you want to review the `src/main/java/com/google/devtools/build/lib/worker/` part?

File: src/main/java/com/google/devtools/build/lib/worker/WorkerMultiplexer.java [hunk 1]
```diff
@@ -150,7 +150,6 @@ public WorkerProcessStatus getStatus() {
   public synchronized void createSandboxedProcess(
       Path workDir, Set<PathFragment> workerFiles, SandboxInputs inputFiles)
       throws IOException, InterruptedException {
-    // TODO: Make blaze clean remove the workdir.
```
Comment (wilwell) [inline src/main/java/com/google/devtools/build/lib/worker/WorkerMultiplexer.java:153]: Why have you removed this line? Is it already cleaning?

**Review by wilwell — CHANGES_REQUESTED** [review]

File: src/main/java/com/google/devtools/build/lib/worker/WorkerMultiplexer.java (hunk 1 above)
Comment (fmeum) [reply src/main/java/com/google/devtools/build/lib/worker/WorkerMultiplexer.java:153]: I took this from @larsrc-google's CL: src/main/java/com/google/devtools/build/lib/worker/WorkerMultiplexer.java

I guess that with multiplex sandboxing Bazel is responsible for cleaning up the work dirs? Not sure though, I can just leave in the comment if we are uncertain about it.

Comment (fmeum) [reply src/main/java/com/google/devtools/build/lib/worker/WorkerMultiplexer.java]: I added it back to be on the safe side.

**Review by wilwell — APPROVED** [review]
Thanks for your contribution!

**Comment by fmeum** [comment]
@wilwell Do you have an opinion regarding https://github.com/bazelbuild/bazel/pull/21370#issuecomment-1946410989?

**Comment by wilwell** [comment]
@fmeum ,  could you explain again  https://github.com/bazelbuild/bazel/pull/21370#issuecomment-1946410989 please?


If I understood correctly you have an issue with  flag `--modify_execution_info` because it only adds the flag key and not flag value (set it to `""` by default).

And what is you suggestion? You want to change the behaviour of [supportsMultiplexSandboxing](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/actions/Spawns.java;bpv=1;bpt=1;l=108?gsn=supportsMultiplexSandboxing&gs=KYTHE%3A%2F%2Fkythe%3A%2F%2Fgithub.com%2Fbazelbuild%2Fbazel%3Flang%3Djava%3Fpath%3Dcom.google.devtools.build.lib.actions.Spawns%23fd5142c38d59e275219d684af419e371104f13ea744d5bba4ecf2cf2aa651e65)?

**Comment by fmeum** [comment]
@wilwell Yes, that would be one way of achieving my goal: I would like to see *some* way to enable sandbox multiplexing for Java actions that 1) doesn't require building a custom Bazel binary and 2) is considered experimental enough that you are willing to accept it before committing to fully support the feature. An alternative could be to expose this on `java_toolchain` as in Lars' CL, but without enabling it for the default toolchains. If you see a different way that you like better, I'm of course open to that as well.

**Comment by wilwell** [comment]
Ahh, sorry I understood your alternative. Yes, let's try add toolchain flag `javac_supports_worker_multiplex_sandboxing` as Lars did in his CL. 

My thoughts about alternatives

1) Change `supports-multiplex-sandboxing` check to compare with `""`. It makes check on worker flags inconsistent and could confuse engineers in future.
2)  Change the behaviour of `modify_execution_info ` flag to accept key-value inputs. Its more reliable solution, but it need the large change in  `modify_execution_info` flag and checks that it will not ruin other builds, which is more complicated.
3) Create `javac_supports_worker_multiplex_sandboxing` toolchain flag and set it `false` by default. With this solution engineers who want to interact multiplex Java workers  could do it by one flag set. It doesn't require large change in code. The only weakness which I see is inefficient use of remote cache, because the toolchain will be different, if I remember correctly.

@fmeum  Am I right, that if you add `javac_supports_worker_multiplex_sandboxing` flag with default `false` value one could override this value for their build in their `bazelrc` file? What is this change?

**Comment by fmeum** [comment]
@wilwell Done, please review again. I also added a test.

File: src/main/java/com/google/devtools/build/lib/worker/WorkerMultiplexer.java (hunk 1 above)
Comment (wilwell) [reply src/main/java/com/google/devtools/build/lib/worker/WorkerMultiplexer.java:153]: Could you please add this comment back?

**Review by wilwell — APPROVED** [review]
Looks good to me.

File: src/main/java/com/google/devtools/build/lib/worker/WorkerMultiplexer.java (hunk 1 above)
Comment (fmeum) [reply src/main/java/com/google/devtools/build/lib/worker/WorkerMultiplexer.java]: Done!

**Comment by fmeum** [comment]
@wilwell Could you add the `awaiting-PR-merge` label in addition to your approval?

**Review by wilwell — APPROVED** [review]

**Comment by fmeum** [comment]
@sgowroji Is this already in the process of being imported?

**Comment by cushon** [comment]
> Is this already in the process of being imported?

There are some internal references to the modified code that require small updates, but it's making progress.

**Comment by iancha1992** [comment]
The changes in this PR have been included in [Bazel 7.1.0 RC2](https://github.com/bazelbuild/bazel/releases/tag/7.1.0rc2). Please test out the release candidate and report any issues as soon as possible.
If you're using Bazelisk, you can point to the latest RC by setting USE_BAZEL_VERSION=7.1.0rc2. Thanks!
